tests/test-host-session.vala

```diff
@@ -2991,35 +2991,104 @@ namespace Frida.HostSessionTest {
 				return;
 			}
 
-			var device_serial = "8B3X1335R";
 			Cancellable? cancellable = null;
+			var device_serial = "99UAY1BUBQ";
+			string debuggable_app = "oversecured.ovaa";
+
+			string gadget_so = "gadget.so";
+			string rpath = "/data/local/tmp/";
+			string lpath = "/tmp/" + gadget_so;
+			string filename = rpath + gadget_so;
+			string mode = "%d".printf (0100666);
+
+			var cmd_buf = new MemoryOutputStream.resizable ();
+			var cmd = new DataOutputStream (cmd_buf);
+			cmd.byte_order = LITTLE_ENDIAN;
 
 			try {
+				yield Frida.Droidy.ShellCommand.run (
+					"am set-debug-app -w --persistent '%s'".printf (debuggable_app), device_serial, cancellable
+				);
+
 				var c = yield Frida.Droidy.Client.open (cancellable);
 
 				c.message.connect ((payload) => {
 					printerr ("Got a message: %s\n", payload);
+					var m = new StringBuilder ();
+					for (var i = 0; i != payload.length; i++) {
+						if (i > 0)
+							m.append (" ");
+						m.append_printf ("%02x", payload[i]);
+					}
+					printerr ("Str: %s\n", m.str);
 				});
 
-				printerr ("A\n");
+				c = yield Frida.Droidy.Client.open (cancellable);
 				yield c.request ("host:transport:" + device_serial, cancellable);
-				printerr ("B\n");
 				yield c.request ("sync:", cancellable);
 
-				// write(3, "SEND\34\0\0\0/data/local/tmp/om.jpg,3"..., 36) = 36
-				var cmd_buf = new MemoryOutputStream.resizable ();
-				var cmd = new DataOutputStream (cmd_buf);
-				cmd.byte_order = LITTLE_ENDIAN;
 				cmd.put_string ("SEND");
-				string filename = "/data/local/tmp/test.bin";
-				string mode = "%d".printf (0100644);
 				cmd.put_uint32 (filename.length + 1 + mode.length);
 				cmd.put_string (filename);
 				cmd.put_string (",");
 				cmd.put_string (mode);
-				printerr ("C\n");
-				yield c.raw_request (cmd_buf.steal_as_bytes (), ACK, cancellable);
-				printerr ("D\n");
+
+				try {
+					File file = File.new_for_path (lpath);
+					Bytes content = file.load_bytes ();
+
+					size_t bytes_read = content.get_size ();
+					size_t MAX_DATA_SIZE = 65536;
+					double data_chunks = bytes_read / (float) MAX_DATA_SIZE;
+
+					int chunks = (int) Math.ceil (data_chunks);
+
+					if (chunks > 0) {
+						int index = 0;
+						size_t bytes_written = 0;
+						size_t written = 0;
+						size_t remaining = 0;
+						size_t end = MAX_DATA_SIZE;
+
+						create_adb_data_payload_header (cmd, chunks, bytes_read);
+
+						while (bytes_written < bytes_read && chunks > 0) {
+							if (chunks == 1)
+								remaining = bytes_read - bytes_written;
+
+							if (remaining > 0 && remaining < MAX_DATA_SIZE)
+								end = remaining;
```

Comment:
```suggestion
							if (chunks == 1)
								remaining = bytes_read - bytes_written;

							if (remaining > 0 && remaining < MAX_DATA_SIZE)
								end = remaining;
```

Feels like this could be simplified. Perhaps `remaining` could start out with the full size and inside the loop one only needs to compute `chunk_size = Math.min (remaining, MAX_DATA_SIZE)`.

Reply:
Took your advice, ready to be reviewed @oleavr